### ISSUE TITLE
kata-collect-data.sh fails after manual installation of kata v2.0.0

### ISSUE THREAD

**Issue by cskarby** [kata-containers/kata-containers#1164]
# Description of problem

1. Install CentOS 8
2. Install [kata v2.0.0](https://github.com/kata-containers/kata-containers/releases/download/2.0.0/kata-static-2.0.0-x86_64.tar.xz) following the [official manual installation instructions](https://github.com/kata-containers/kata-containers/blob/2.0.0/docs/install/container-manager/containerd/containerd-install.md#install-kata-containers)
3. Run `kata-collect-data.sh`

# Expected result

It is expected that `kata-collect-data.sh` should work even after a [manual installation following the official instructions](https://github.com/kata-containers/kata-containers/blob/2.0.0/docs/install/container-manager/containerd/containerd-install.md#install-kata-containers).

# Actual result

```sh
# kata-collect-data.sh 
ERROR: kata-collect-data.sh: cannot find runtime ''
```

# Further information

I've created a container image to easily reproduce the error, see https://github.com/cskarby/kata-v2-centos8-fail

```sh
$ docker run -it cskarby/kata-v2-centos8-fail 
[root@23e50414a5ad /]# export PATH=/opt/kata/bin:$PATH
[root@23e50414a5ad /]# kata-runtime --version
  : 2.0.0
   commit   : aa295c91f2ec5ec3389992111a1c6acecb46024b
   OCI specs: 1.0.1-dev
[root@23e50414a5ad /]# kata-collect-data.sh 
ERROR: kata-collect-data.sh: cannot find runtime ''
[root@23e50414a5ad /]# 
```

**Comment by fidencio** [comment]
Please, could someone from @kata-containers/redhat take a look at this?

**Comment by pmores** [comment]
@fidencio Sure!  The immediate cause seems to be simply that the `runtime_name` variable in `kata-collect-data.sh` in the testing container actually is empty.  Given that the script is built from a template that takes the runtime name as a parameter, I suspect something went wrong during build.

I'll be looking into this some more.  If I can get any more useful info I'll let you know.

**Comment by pmores** [comment]
@cskarby Do I understand correctly that you downloaded a release (as opposed to making your own build)?  If so, which version was it?

Frankly I don't see how this could happen with the current code however it was apparently possible less than two months ago.

**Comment by cskarby** [comment]
@pmores , yes that is correct, https://github.com/cskarby/kata-v2-centos8-fail/blob/e5743617ff7da1702cca87038e0cb9822cca18fb/Dockerfile#L6 downloads the artifact from the kata-containers release v2.0.0, see bottom of https://github.com/kata-containers/kata-containers/releases/tag/2.0.0 which was released October 19th.

**Comment by pmores** [comment]
@cskarby Thanks for the info, it seems to make sense now.  `kata-collect-data.sh` is generated in build time from a template (`kata-collect-data.sh.in`) by replacing variables in the template with actual values.  This replacement was basically broken until [this commit](https://github.com/kata-containers/kata-containers/commit/966bd573442d2f0b951e21aae2445dd009b3d5d1) fixed it for most variables and [this one](https://github.com/kata-containers/kata-containers/commit/96a4ed7d709d1eef31b2421f5a88d8f500583ea0) finished the fix for `RUNTIME_NAME` in particular.

Now the former commit (from Oct 14) made it into the release but the latter one (from Oct 19, the very day the release was made public) did not - this is apparent from the Short Log section of the [release page](https://github.com/kata-containers/kata-containers/releases/tag/2.0.0) that you linked to.  Also, the result you're seeing is consistent with this.  The former of the two commits ensured that variables in templates are replaced with values of the same-named variables in the runtime Makefile.  However, until the latter commit, there was no `RUNTIME_NAME` variable in the Makefile so `RUNTIME_NAME` in `kata-collect-data.sh.in` was likely replaced by an empty string.  This is where the `cannot find runtime ''` part of your output comes from.

All in all, the release you're using is buggy but the bug has already been fixed (I verified this both by code inspection and a test build so hopefully I'm not mistaken).  Given this, I'd say your options at this moment are either build your own runtime from latest HEAD, or, if you can't do that, wait for a next release.  I don't know when next release is coming but @fidencio might know or might at least know who knows. :-)

@fidencio It might make sense for us to consider doing a bugfix release IMO, seeing as if we'd like people to download 2.0 and report problems, it's in our best interest for them to have a working `kata-collect-data.sh`. ;-)

@cskarby Thanks for a very cool bug report - I mean wow, including a container with a repro case makes it really easy to track down stuff!! :-)

**Comment by cskarby** [comment]
@pmores thank you for the feedback. It made sense to put it in a container, as this problem was reproducible even without starting any VM. As the problem was in the released artifact and had nothing to do with CentOS 8, I've removed the distro info from the name of this bug report.

Setting runtime_name to kata-runtime makes it possible to run `kata-collect-data.sh`, however I'm not sure if this is a complete fix for the static assets from v2.0.0, as the `@RUNTIME_NAME@` variable may have been in other files as well?
```diff
--- opt/kata/bin/kata-collect-data.sh	2020-12-03 14:42:40.768781729 +0000
+++ opt/kata/bin/kata-collect-data.sh	2020-12-03 14:43:03.846780118 +0000
@@ -6,7 +6,7 @@
 #
 
 typeset -r script_name=${0##*/}
-typeset -r runtime_name=""
+typeset -r runtime_name="kata-runtime"
 typeset -r runtime_path=$(command -v "$runtime_name" 2>/dev/null)
 typeset -r runtime_snap_name="kata-containers.runtime"
 typeset -r runtime_snap_path=$(command -v "$runtime_snap_name" 2>/dev/null)
```

I think it would sense to release a v2.0.1 with 96a4ed7d709d1eef31b2421f5a88d8f500583ea0 to fix this bug :+1:

**Comment by pmores** [comment]
@cskarby You're probably right, there might be a third option - patching the release by hand. :-)  Well as far as I can see, there's one other place in `kata-collect-data.sh` that you need to change which is `find_system_journal_problems()` function invocation in `show_runtime_log_details()` (should be somewhere around the line 309) whose second argument should be `@RUNTIME_NAME@` but you'll probably see an empty string there.  Without this the script will probably crash trying to show its output - I think it will basically try to run `journalctl -t` without an argument (I haven't verified this though in practice).

However, fixing the above might not actually help all that much as the last `RUNTIME_NAME` reference that I can see is in a `.go` file (`cli/config-generated.go`) where a variable called `name` is set to `@RUNTIME_NAME@` value.  I can imagine this variable being used to tag log messages in which case the above change to `kata-collect-data.sh` would probably just prevent the script from crashing but its output would still be incomplete (it wouldn't find kata-runtime's log messages).  To fix this, I think at this point you'd be better off if you do your own build.

**Comment by fidencio** [comment]
@pmores, @cskarby, what's the outcome of this issue?

Is this that https://github.com/kata-containers/kata-containers/commit/96a4ed7d709d1eef31b2421f5a88d8f500583ea0 should be backported to stable-2.0?  If so, @c3d, could you do that?

**Comment by c3d** [comment]
@fidencio Sorry for missing that one. Here is the PR: https://github.com/kata-containers/kata-containers/pull/1762

**Comment by snir911** [comment]
This one should have been closed by #1762
@c3d please reopen if I missed something
